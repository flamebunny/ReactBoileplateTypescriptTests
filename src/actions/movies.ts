import { fetchJsonFromApi, MoviesProps, getPostParameters } from '../api'

export const GET_MOVIES = 'GET_MOVIES'
export function getMovies(movies) {
  return {
    type:  GET_MOVIES,
    movies: [{ title: 'Lord of The Rings' }, { title: 'The Matrix' }]
  }
}

export const MOVIES_FETCH = 'MOVIES_FETCH'
export function moviesFetch() {
  return {
    type: MOVIES_FETCH
  }
}

export interface GetSuperPicksRequest {
  FixtureDate: string
  FixtureId: string
  RaceNumber: number
  StarterNumber: number
  PropositionSeq?: string
}

export function getMovieList(request: GetSuperPicksRequest): Promise<MoviesProps> {
  return fetchJsonFromApi<MoviesProps>('https://facebook.github.io/react-native/movies.json', getPostParameters(request))
    .catch(async error => {
      const errorResponse = await error.response.json()
      throw {
        response: errorResponse
      }
    })
}

export const MOVIES_FETCH_ASYNC = 'MOVIES_FETCH_ASYNC'
export function moviesFetchAsync() {

  const request: GetSuperPicksRequest = {
    FixtureDate: '',
    FixtureId: '',
    RaceNumber: 0,
    StarterNumber: 0,
    PropositionSeq: 'test' || null
  }

  return (dispatch, getStore) => {
    if (getStore().fetching) {
      return
    }
    console.log('bbbb')
    dispatch(moviesFetch())
    console.log('cccc')
    getMovieList(request)
      .then(response => dispatch(moviesFetchSuccess(response)))
      .catch(error => dispatch(moviesFetchFailed({ error })))
  }

}

export const MOVIES_FETCH_SUCCESS = 'MOVIES_FETCH_SUCCESS'
export function moviesFetchSuccess({ movies }) {
  return {
    type: MOVIES_FETCH_SUCCESS,
    movies
  }
}

export const MOVIES_FETCH_FAILED = 'MOVIES_FETCH_FAILED'
export function moviesFetchFailed({ error }) {
  return {
    type: MOVIES_FETCH_FAILED,
    errorMessage: error.message
  }
}
