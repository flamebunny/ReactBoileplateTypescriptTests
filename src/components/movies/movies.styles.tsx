import styled from 'react-emotion'

export const LoadingStyled = styled('div')(
  {
    background: '#EEE'
  }
)
