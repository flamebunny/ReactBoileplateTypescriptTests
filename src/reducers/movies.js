const defaultState = {
  fetching: false,
  movies: [],
  errorMessage: ''
}

export function moviesReducer(state = defaultState, action) {
  switch (action.type) {
    case 'GET_MOVIES':
      return {
        ...state,
        movies: action.movies
      }

    case 'MOVIES_FETCH':
      return {
        ...state,
        fetching: true
      }

    case 'MOVIES_FETCH_SUCCESS':
      return {
        ...state,
        fetching: false,
        errorMessage: '',
        movies: action.movies
      }

    case 'MOVIES_FETCH_FAILED':
      return {
        ...state,
        fetching: false,
        errorMessage: action.errorMessage
      }

    default:
      return state
  }
}