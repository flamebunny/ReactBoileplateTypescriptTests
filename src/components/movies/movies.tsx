import * as React from 'react'
import { connect } from 'react-redux'
import { Movie } from '../movie/movie'
import { LoadingStyled } from './movies.styles'

export const Movies = ({ fetching, movies }) => {
  return (
    <div>
      {fetching && <LoadingStyled>Loading Movies</LoadingStyled>}
      Awesome
      {movies && movies.map(movie => <Movie key={movie.id} {...movie} />)}
    </div>
  )
}

export default connect(state => state.movies)(Movies)
