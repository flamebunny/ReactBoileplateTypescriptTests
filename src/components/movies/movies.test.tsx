import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import 'jest-enzyme'
import { Movies } from './movies'
import { LoadingStyled } from './movies.styles'
import { Movie } from '../movie/movie'

interface MovieListProps {
  id: number
  title: string
  releaseYear: number
}

const movies: MovieListProps[] = [
  { id: 1, title: 'Star Wars', releaseYear: 1977 },
  { id: 2, title: 'Back to the Future', releaseYear: 1985 },
  { id: 3, title: 'The Matrix', releaseYear: 1999 },
  { id: 4, title: 'Inception', releaseYear: 2010 },
  { id: 5, title: 'Interstellar', releaseYear: 2014 }
]

describe('The Movies component', () => {
  describe('When rendered', () => {
    let component: ReactWrapper

    beforeEach(() => {
      component = mount(<Movies movies={movies} fetching={false} />)
    })

    afterEach(() => {
      if (component) {
        component.unmount()
        jest.resetAllMocks()
      }
    })

    it('Matches the snapshot', () => {
      expect(component).toMatchSnapshot()
    })

  })

  describe('When fetching is true', () => {
    describe('When rendered', () => {
      let component: ReactWrapper

      beforeEach(() => {
        component = mount(<Movies movies={[]} fetching={true} />)
      })

      afterEach(() => {
        if (component) {
          component.unmount()
          jest.resetAllMocks()
        }
      })

      it('Renders loading signal', () => {
        expect(component.find(LoadingStyled).text()).toBe('Loading Movies')
      })

    })
  })

  describe('When Movies are available', () => {
    describe('When rendered', () => {
      let component: ReactWrapper

      beforeEach(() => {
        component = mount(<Movies movies={movies} fetching={false} />)
      })

      afterEach(() => {
        if (component) {
          component.unmount()
          jest.resetAllMocks()
        }
      })

      it('Display 5 movie list items', () => {
        expect(component.find(Movie).length).toEqual(5)
      })

    })
  })
})
