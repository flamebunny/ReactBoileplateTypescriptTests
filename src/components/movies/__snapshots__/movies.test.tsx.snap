// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Movies component When rendered Matches the snapshot 1`] = `
<Component
  fetching={false}
  movies={
    Array [
      Object {
        "id": 1,
        "releaseYear": 1977,
        "title": "Star Wars",
      },
      Object {
        "id": 2,
        "releaseYear": 1985,
        "title": "Back to the Future",
      },
      Object {
        "id": 3,
        "releaseYear": 1999,
        "title": "The Matrix",
      },
      Object {
        "id": 4,
        "releaseYear": 2010,
        "title": "Inception",
      },
      Object {
        "id": 5,
        "releaseYear": 2014,
        "title": "Interstellar",
      },
    ]
  }
>
  <div>
    <Component
      id={1}
      key="1"
      releaseYear={1977}
      title="Star Wars"
    >
      <div>
        Movie: 
        Star Wars
         (
        1977
        )
      </div>
    </Component>
    <Component
      id={2}
      key="2"
      releaseYear={1985}
      title="Back to the Future"
    >
      <div>
        Movie: 
        Back to the Future
         (
        1985
        )
      </div>
    </Component>
    <Component
      id={3}
      key="3"
      releaseYear={1999}
      title="The Matrix"
    >
      <div>
        Movie: 
        The Matrix
         (
        1999
        )
      </div>
    </Component>
    <Component
      id={4}
      key="4"
      releaseYear={2010}
      title="Inception"
    >
      <div>
        Movie: 
        Inception
         (
        2010
        )
      </div>
    </Component>
    <Component
      id={5}
      key="5"
      releaseYear={2014}
      title="Interstellar"
    >
      <div>
        Movie: 
        Interstellar
         (
        2014
        )
      </div>
    </Component>
  </div>
</Component>
`;
